### PR TITLE
add rcnn ang mask rcnn dygraph

### PR DESCRIPTION
**Pull request by FDInSky**
add rcnn ang mask rcnn dygraph
the model implementation mainly in dir "models/dyg/" and use the train_dyg.py to train.

### CODE REVIEW

**Review by qingqing01 — COMMENTED** [review]
Please clean code at first

**Comment by FDInSky** [comment]
> Please clean code at first
已经清理

File: dygraph/rcnn/README.md [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+[Faster RCNN](https://arxiv.org/abs/1506.01497) 整体网络可以分为4个主要内容：
+
+1. 基础卷积层。作为一种卷积神经网络目标检测方法，Faster RCNN首先使用一组基础的卷积网络提取图像的特征图。特征图被后续RPN层和全连接层共享。本示例采用[ResNet-50](https://arxiv.org/abs/1512.03385)作为基础卷积层。
+2. 区域生成网络(RPN)。RPN网络用于生成候选区域(proposals)。该层通过一组固定的尺寸和比例得到一组锚点(anchors), 通过softmax判断锚点属于前景或者背景，再利用区域回归修正锚点从而获得精确的候选区域。
+3. RoI Align。该层收集输入的特征图和候选区域，将候选区域映射到特征图中并池化为统一大小的区域特征图，送入全连接层判定目标类别, 该层可选用RoIPool和RoIAlign两种方式，在config.py中设置roi\_func。
+4. 检测层。利用区域特征图计算候选区域的类别，同时再次通过区域回归获得检测框最终的精确位置。
+
+[Mask RCNN](https://arxiv.org/abs/1703.06870) 扩展自Faster RCNN，是经典的实例分割模型。
+
+Mask RCNN同样为两阶段框架，第一阶段扫描图像生成候选框；第二阶段根据候选框得到分类结果，边界框，同时在原有Faster RCNN模型基础上添加分割分支，得到掩码结果，实现了掩码和类别预测关系的解藕。
```
Comment (littletomatodonkey) [inline dygraph/rcnn/README.md]: 解藕->解耦

File: dygraph/rcnn/README.md [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+├── val2017
+│   ├── 000000000139.jpg
+│   ├── 000000000285.jpg
+|   ...
+
+```
+
+## 模型训练
+
+**下载预训练模型：** 本示例提供Resnet-50预训练模型，该模性转换自Caffe，并对批标准化层(Batch Normalization Layer)进行参数融合。采用如下命令下载预训练模型：
```
Comment (littletomatodonkey) [inline dygraph/rcnn/README.md]: 模性->模型

File: dygraph/rcnn/README.md [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+    sh ./pretrained/download.sh
+
+通过初始化`pretrained_model` 加载预训练模型。同时在参数微调时也采用该设置加载已训练模型。
+请在训练前确认预训练模型下载与加载正确，否则训练过程中损失可能会出现NAN。
+
+**安装[cocoapi](https://github.com/cocodataset/cocoapi)：**
+
+训练前需要首先下载[cocoapi](https://github.com/cocodataset/cocoapi)：
+
+    git clone https://github.com/cocodataset/cocoapi.git
```
Comment (littletomatodonkey) [inline dygraph/rcnn/README.md:76]: 直接`pip install pycocoapi`会不会更简洁一些呢?

File: dygraph/rcnn/README.md [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+       --model_save_dir=output/ \
+       --pretrained_model=${path_to_pretrain_model} \
+       --data_dir=${path_to_data} \
+       --MASK_ON=True\
+       --roi_func=RoIAlign
+    ```
+
+    - 通过设置export CUDA\_VISIBLE\_DEVICES=0,1,2,3,4,5,6,7指定8卡GPU训练。
+    - 通过设置```MASK_ON```选择Faster RCNN和Mask RCNN模型。
+    - 使用Windows GPU环境的用户，需要设置```parallel```为False，将[fluid.ParallelExecutor](http://paddlepaddle.org/documentation/docs/zh/1.4/api_cn/fluid_cn.html#parallelexecutor)替换为[fluid.Executor](http://paddlepaddle.org/documentation/docs/zh/1.4/api_cn/fluid_cn.html#executor)。
```
Comment (littletomatodonkey) [inline dygraph/rcnn/README.md:110]: 是否考虑换成1.6或者develop的文档链接呢？

File: dygraph/rcnn/box_utils.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+def clip_xyxy_to_image(x1, y1, x2, y2, height, width):
+    """Clip coordinates to an image with the given height and width."""
+    x1 = np.minimum(width - 1., np.maximum(0., x1))
+    y1 = np.minimum(height - 1., np.maximum(0., y1))
+    x2 = np.minimum(width - 1., np.maximum(0., x2))
+    y2 = np.minimum(height - 1., np.maximum(0., y2))
+    return x1, y1, x2, y2
+
+
+def nms(dets, thresh):
```
Comment (littletomatodonkey) [inline dygraph/rcnn/box_utils.py:73]: 是否需要考虑坐标为normalized的情况呢？

File: dygraph/rcnn/eval_helper.py [hunk 6]
```diff
@@ -0,0 +1,389 @@
+#  Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserve.
```
Comment (littletomatodonkey) [inline dygraph/rcnn/eval_helper.py]: 年份需要修改不？

File: dygraph/rcnn/infer.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
+import os
```
Comment (littletomatodonkey) [inline dygraph/rcnn/infer.py]: copyright？

File: dygraph/rcnn/models/dyg/cyops/1 [hunk 8]
```diff
@@ -0,0 +1,285 @@
+cimport cython
```
Comment (littletomatodonkey) [inline dygraph/rcnn/models/dyg/cyops/1]: should be deleted?

File: dygraph/rcnn/box_utils.py [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import numpy as np
+
+
+def xywh_to_xyxy(xywh):
```
Comment (littletomatodonkey) [inline dygraph/rcnn/box_utils.py:31]: 部分函数是否可以和models/dyg/pyops/bbox.py合并呢？

File: dygraph/rcnn/.run_ce.sh [hunk 10]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
```
Comment (qingqing01) [inline dygraph/rcnn/.run_ce.sh]: please remove CE in dygraph

File: dygraph/rcnn/README.md [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+下表为模型评估结果：
+
+Faster RCNN
+
+| 模型                   |   RoI处理方式  | 批量大小   | 迭代次数   | mAP  |
+| :--------------- | :--------: | :------------:    | :------------------:    |------: |
+| [Fluid RoIPool minibatch padding](http://paddlemodels.bj.bcebos.com/faster_rcnn/model_pool_minibatch_padding.tar.gz) | RoIPool | 8   |    180000        | 0.316 |
+| [Fluid RoIPool no padding](http://paddlemodels.bj.bcebos.com/faster_rcnn/model_pool_no_padding.tar.gz)  | RoIPool | 8   |    180000        | 0.318 |
+| [Fluid RoIAlign no padding](http://paddlemodels.bj.bcebos.com/faster_rcnn/model_align_no_padding.tar.gz)  | RoIAlign | 8   |    180000        | 0.348 |
+| [Fluid RoIAlign no padding 2x](http://paddlemodels.bj.bcebos.com/faster_rcnn/model_align_no_padding_2x.tar.gz)  | RoIAlign | 8   |    360000        | 0.367 |
```
Comment (qingqing01) [inline dygraph/rcnn/README.md:170]: 这里的模型能直接加载吗？ 如果不能，请去掉

File: dygraph/rcnn/_ce.py [hunk 12]
```diff
@@ -0,0 +1,62 @@
+# this file is only used for continuous evaluation test!
```
Comment (qingqing01) [inline dygraph/rcnn/_ce.py]: please remove CE in this PR

File: dygraph/rcnn/eval_dyg.py [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+            with io.open("detection_segms_result.json", 'w') as outfile:
+                encode_func = unicode if six.PY2 else str
+                outfile.write(encode_func(json.dumps(segms_res)))
+            print("start evaluate mask using coco api")
+            cocoDt = cocoGt.loadRes("detection_segms_result.json")
+            cocoEval = COCOeval(cocoGt, cocoDt, 'segm')
+            cocoEval.evaluate()
+            cocoEval.accumulate()
+
+    eval_loop()
```
Comment (qingqing01) [inline dygraph/rcnn/eval_dyg.py]: 直接展开代码就好了，没必要加个eval_loop()的空函数

File: dygraph/rcnn/learning_rate.py [hunk 14]
```diff
@@ -0,0 +1,61 @@
+#  Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserve.
```
Comment (qingqing01) [inline dygraph/rcnn/learning_rate.py]: 这个文件动态图也用了吗?

File: dygraph/rcnn/models/dyg/pyops/roi_extractor.py [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+
+def roi_pool(input_x, rois, pooled_height, pooled_width, spatial_scale):
+    input_x = input_x.numpy()
+    rois = rois.numpy()
+    batch_size, channels, height, width = input_x.shape
+    print("debug roi pool")
+    print("debug input feat: ", input_x.shape)
+    rois_num = rois.shape[1]
+    #out_data = np.zeros((rois_num, channels, pooled_height, pooled_width))
+    #argmax_data = np.zeros((rois_num, channels, pooled_height, pooled_width))
```
Comment (qingqing01) [inline dygraph/rcnn/models/dyg/pyops/roi_extractor.py]: remove line 14 and line 15

File: dygraph/rcnn/models/model_builder.py [hunk 16]
```diff
@@ -0,0 +1,441 @@
+#  Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserve.
```
Comment (qingqing01) [inline dygraph/rcnn/models/model_builder.py:1]: 去掉动态图不用的静态图配置

**Review by qingqing01 — COMMENTED** [review]
还有很多静态图的配置在里面，觉得还是删除比较好~

File: dygraph/rcnn/models/model_builder.py (hunk 16 above)
Comment (FDInSky) [reply dygraph/rcnn/models/model_builder.py:1]: eval时，目前需要转成静态来做，最新的develop的动态图，显存爆掉

File: dygraph/rcnn/models/dyg/pyops/roi_extractor.py (hunk 15 above)
Comment (FDInSky) [reply dygraph/rcnn/models/dyg/pyops/roi_extractor.py]: 已删

File: dygraph/rcnn/learning_rate.py (hunk 14 above)
Comment (FDInSky) [reply dygraph/rcnn/learning_rate.py]: 已删

File: dygraph/rcnn/eval_dyg.py (hunk 13 above)
Comment (FDInSky) [reply dygraph/rcnn/eval_dyg.py]: 已改

File: dygraph/rcnn/_ce.py (hunk 12 above)
Comment (FDInSky) [reply dygraph/rcnn/_ce.py]: 已删

File: dygraph/rcnn/.run_ce.sh (hunk 10 above)
Comment (FDInSky) [reply dygraph/rcnn/.run_ce.sh]: 已删

File: dygraph/rcnn/models/dyg/cyops/1 (hunk 8 above)
Comment (FDInSky) [reply dygraph/rcnn/models/dyg/cyops/1]: 已删

File: dygraph/rcnn/eval_helper.py (hunk 6 above)
Comment (FDInSky) [reply dygraph/rcnn/eval_helper.py]: 已修改

File: dygraph/rcnn/README.md (hunk 2 above)
Comment (FDInSky) [reply dygraph/rcnn/README.md]: 已改

File: dygraph/rcnn/README.md (hunk 1 above)
Comment (FDInSky) [reply dygraph/rcnn/README.md]: 已改

File: dygraph/rcnn/infer.py (hunk 7 above)
Comment (FDInSky) [reply dygraph/rcnn/infer.py]: 已删

**Review by qingqing01 — APPROVED** [review]
